src/util/pushSender.ts

```diff
@@ -1,120 +1,118 @@
-    if (apiKeyRow == null || apiKeyRow.adminsdk == null) {
-      senders.set(apiKey, null)
-      return null
-    }
-
-    // TODO: We have never passed the correct data type here,
-    // so either update our database or write a translation layer:
-    const serviceAccount: any = apiKeyRow.adminsdk
-    const projectId = serviceAccount.projectId ?? serviceAccount.project_id
-
-    // Create a sender if we have an API key for them.
-    // It is possible that multiple API keys will use the same
-    // Firebase project, so check for existing instances
-    // before creating new ones:
-    const app =
-      admin.apps.find(app => app?.name === projectId) ??
-      admin.initializeApp(
-        { credential: admin.credential.cert(serviceAccount) },
-        projectId
-      )
-    const sender = app.messaging()
-    senders.set(apiKey, sender)
-    return sender
-  }
+  // Persistent delivery:
+  const deliveryMode = 2
 
   async function sendToEvent(
     event: PushEvent,
-    message: SendableMessage
+    message: SendableMessage,
+    priority?: number
   ): Promise<void> {
     const { deviceId, loginId } = event
-
-    if (loginId != null) {
-      await sendToLogin(loginId, message)
-    } else if (deviceId != null) {
-      const device = await getDeviceById(connections, deviceId, new Date())
-      await sendToDevice(device.device, message)
-    }
+    await connections.queue.publish(
+      wasRabbitMessage({
+        deviceId,
+        loginId,
+        message
+      }),
+      { deliveryMode, priority }
+    )
   }
 
   async function sendToLogin(
     loginId: Uint8Array,
-    message: SendableMessage
+    message: SendableMessage,
+    priority?: number
   ): Promise<void> {
-    const deviceRows = await getDevicesByLoginId(connections, loginId)
-    for (const deviceRow of deviceRows) {
-      await sendToDevice(deviceRow.device, message)
-    }
+    await connections.queue.publish(
+      wasRabbitMessage({
+        deviceId: undefined,
+        loginId,
+        message
+      }),
+      { deliveryMode, priority }
+    )
   }
 
   async function sendToDevice(
     device: Device,
-    message: SendableMessage
+    message: SendableMessage,
+    priority?: number
   ): Promise<void> {
-    const { apiKey, deviceToken } = device
-
-    if (message.isPriceChange && device.ignorePriceChanges) return
-    if (apiKey == null || deviceToken == null) return
-
-    const sender = await getSender(apiKey)
-    if (sender == null) return
-
-    try {
-      await sender.send({
-        token: deviceToken,
-        notification: {
-          title: message.title ?? '',
-          body: message.body ?? ''
-        },
-        data: message.data ?? {}
-      })
-    } catch (err) {}
+    await connections.queue.publish(
+      wasRabbitMessage({
+        deviceId: device.deviceId,
+        loginId: undefined,
+        message
+      }),
+      { deliveryMode, priority }
+    )
   }
 
-  const instance: PushSender = {
```

Comment:
One benefit to an `instance` identifier is that the red-squiglies will isolate to this declaration instead of the entire function body. Optionall, keep it.

Reply:
Done